src/commands/release-asset/download.ts

```diff
@@ -0,0 +1,145 @@
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import { Command, Flags } from '@oclif/core';
+import { getBalenaSdk, stripIndent, getVisuals } from '../../utils/lazy';
+import { commitOrIdArg } from '../release';
+import * as fs from 'fs';
+import * as path from 'path';
+import { pipeline } from 'stream/promises';
+import { Transform } from 'stream';
+
+export default class ReleaseAssetDownloadCmd extends Command {
+	public static description = stripIndent`
+		Download a release asset.
+
+		Download a release asset with the specified key. By default, the file will be saved
+		with the original filename. Use the --output flag to specify a different output path.
+	`;
+
+	public static examples = [
+		'$ balena release-asset download 1234567 --key config.json',
+		'$ balena release-asset download a777f7345fe3d655c1c981aa642e5555 --key app.tar.gz --output ./downloads/app.tar.gz',
+		'$ balena release-asset download 1234567 --key large-file.bin -o output.bin',
+	];
+
+	public static args = {
+		commitOrId: commitOrIdArg({
+			description: 'the commit or ID of the release',
+			required: true,
+		}),
+	};
+
+	public static flags = {
+		key: Flags.string({
+			description: 'the key of the release asset to download',
+			required: true,
+		}),
+		output: Flags.string({
+			char: 'o',
+			description: 'output path for the downloaded file',
+		}),
+	};
+
+	public static authenticated = true;
+
+	public async run() {
+		const { args, flags } = await this.parse(ReleaseAssetDownloadCmd);
+		const balena = getBalenaSdk();
+
+		const release = await balena.models.release.get(args.commitOrId, {
+			$select: ['id'],
+		});
+
+		const releaseAsset = await balena.models.release.asset.get(
+			{
+				release: release.id,
+				asset_key: flags.key,
+			},
+			{
+				$select: ['id', 'asset'],
+			},
+		);
+
+		if (releaseAsset.asset == null) {
+			throw new Error(`Release asset ${releaseAsset.id} is empty`);
+		}
+
+		const stream = await balena.models.release.asset.download({
+			release: release.id,
+			asset_key: flags.key,
+		});
+
+		let outputPath: string;
+		if (flags.output) {
+			outputPath = path.resolve(flags.output);
+		} else {
+			// Use the asset key as the filename since we don't have headers
+			outputPath = path.resolve(flags.key);
+		}
+
+		const dir = path.dirname(outputPath);
+		await fs.promises.mkdir(dir, { recursive: true });
+
+		const visuals = getVisuals();
+		const bar = new visuals.Progress(
+			`Downloading release asset '${flags.key}'`,
+		);
+		const spinner = new visuals.Spinner(
+			`Downloading release asset '${flags.key}' (size unknown)`,
```

Comment:
Again I question the intent and usefulness of a constant `size unknown`

Reply:
Especially as it appears we have the size below. But I also don't expect people even want to see the total size they are downloading, but a fraction indicating how much of the total size they have downloaded now (i.e. 250MB/3GB)

Reply:
> But I also don't expect people even want to see the total size they are downloading, but a fraction indicating how much of the total size they have downloaded now (i.e. 250MB/3GB)

That is of course what the progress bar is for, so again I wonder if there's any point even showing a size here

Reply:
As I mentioned before, I am removing the spinner altogether :+1: